### PR TITLE
IPv6 containers and NAT

### PR DESCRIPTION
**Pull request by valeriangalliat**
Following #4917, I updated the containers and NAT modules to support IPv6:
- Allow to define an IPv6 host and local address for a container.
- Support IPv6 port forwarding in NAT configuration.

Here's an example `configuration.nix` I use with these features:

``` nix
{ config, pkgs, ... }:

{
  networking = {
    firewall.allowedTCPPorts = [ 80 ];

    nat = {
      enable = true;
      internalInterfaces = ["ve-+"];
      externalInterface = "enp0s3";

      forwardPorts = [
        { sourcePort = 80; destination = "192.168.100.2:80"; }
      ];

      forwardIPv6Ports = [
        { sourcePort = 80; destination = "[fc00::2]:80"; }
      ];
    };
  };

  containers = {
    web = {
      autoStart = true;

      privateNetwork = true;
      hostAddress = "192.168.100.1";
      localAddress = "192.168.100.2";
      hostIPv6Address = "fc00::1";
      localIPv6Address = "fc00::2";

      config = { config, pkgs, ... }: {
        networking.firewall.allowedTCPPorts = [ 80 443 ];

        services.httpd = {
          enable = true;
          adminAddr = "admin@example.com";
        };
      };
    };
  };
}
```

I'm not sure about the options naming, sometimes it's `IPv6`, sometimes `ipv6` or `ip6`, or just adding `6` to the option name (`defaultGateway6`).

There's also a lot of copy/paste with only `s/iptables/ip6tables/` in `nat.nix`, I'd appreciate input on how to factorize this.
###### Todo
- [x] Make this feature optional, according to `networking.enableIPv6` and a new option `networking.nat.enableIPv6`.
- [x] Unify `forwardIPv6Ports` and `forwardPorts` (determine whether to call iptables or ip6tables depending on the destination address).

### CODE REVIEW

**Comment by bluescreen303** [comment]
why would you NAT IPv6?
Even end-user subscribers receive /64 subnets

**Comment by valeriangalliat** [comment]
I use containers to manage my services internally, but I want them to appear on the same public IP (both v4 and v6), that is the main server IP. That's why I use IPv6 NAT.

Lennart Pottering explains the point in a comment on [this post](https://plus.google.com/+LennartPoetteringTheOneAndOnly/posts/TJXfHmMUmbM), I just copy the comment here since I can't link to a specific comment on Google+:

> Well, if you consider nspawn containers less "autonomous machines on their own" and more "appliance sandboxes" then it might very well make sense to deploy IPv6 NAT for this, so that for the outside everything appears to be coming from the same host, instead of dumping onto the outside via the address namespace how the system is internally partitioned into containers.

**Comment by lethalman** [comment]
@wkennington what do you think about this? I never used ip6tables otherwise it's :+1: for me

**Comment by valeriangalliat** [comment]
BTW this IPv6 NAT is maybe an edge case, do you think it should be disabled by default? A configuration option like `networking.firewall.nat.ipv6 = true;` to enable it explicitely (and maybe enable it automatically if `forwardIPv6Ports` is set)? I don't want to bloat everyone's nat with this IPv6 feature if it's not needed.

**Comment by lethalman** [comment]
@valeriangalliat yes that's a good idea. Also && with http://nixos.org/nixos/manual/ch-options.html#opt-networking.enableIPv6

**Comment by bluescreen303** [comment]
@valeriangalliat I understand the use case (and poetering's explanation) and I'm not against this feature. I'm just worried people will stick to some oldschool thinking and not use what ipv6 gives us :)

Personally, for servers, I tend not to think about "the main ip". Instead, I configure services to each get their own private ip. That way, the outside still has no clue about system internals (whether there's 1 server or many, whether services run in a container or not) and I found stuff like traffic shaping and firewalling to become much easier.

The "applicance sandbox" case is served very well by just handing 1 application a personal ip address. No need to worry about exactly which ports are used and how and check if this changes on upgrades. Firewalling and traffic shaping really becomes per-application :)

**Comment by lethalman** [comment]
@bluescreen303 I don't think that's the point really. You could have a cost with setting up dns entries etc., yet you really want a single IP exposed.

Rather the point is: why would you need ipv6 for NAT where you can use ipv4 for internal IPs? Don't tell me you saturated all the 192.168.x.x and 10.x.x.x IPs :)

**Comment by valeriangalliat** [comment]
@bluescreen303 I'm not sure to understand, you're suggesting each container have a public IPv6 address, and setup DNS entries to address each container? It's probably the best solution to embrace IPv6, but in a first place I wanted to avoid setting up DNS entries for each container, and rather centralize everything on the host IP.

@lethalman I could indeed stick to IPv4 for internal IPs, NAT64 would be a better solution but I didn't managed to get it working the last time I tried. I'll give it a go again by tomorrow. It would be handy to have a way to configure NAT64 from NixOS configuration anyway!

**Comment by lethalman** [comment]
@valeriangalliat btw for me the PR is ok as long as it's optional

File: nixos/modules/services/networking/nat.nix [hunk 1]
```diff
@@ -99,6 +131,18 @@ in
         '';
     };
 
+    networking.nat.internalIPs6 = mkOption {
```
Comment (edolstra) [inline nixos/modules/services/networking/nat.nix]: `internalIPs6` looks weird. I would call it `internalIPv6s` or`internalIPv6Ranges`.

**Comment by edolstra** [comment]
Does `forwardIPv6Ports` allow forwarding local IPv4 ports to an IPv6 address?

Maybe `forwardIPv6Ports` and `forwardPorts` can be unified. The only way they differ is in the destination...

**Comment by valeriangalliat** [comment]
> Does `forwardIPv6Ports` allow forwarding local IPv4 ports to an IPv6 address?

It's not the case currently; ip6tables don't support DNAT on an IPv4 address, we need to use iptables for this. However you're right, the configuration on Nix side could be unified, we just need to figure out whether the destination is an IPv4 or IPv6 address and call either ip6tables or iptables.

I see two solutions for this: `iptables $args || ip6tables $args`; if iptables fails, it's probably because it didn't understood the IPv6 address, so try with ip6tables. Don't looks really neat, but it could work. Another solution is to try to determine ourselves if the destination is IPv6 or IPv4, for exemple by counting the `:` occurrences (`<= 1` means IPv4, otherwise IPv6), and call ip6tables or iptables accordingly.

---

Also it looks like [ip6tables don't support port forwarding on an IPv4 address (NAT64)](http://serverfault.com/questions/276515/use-iptables-to-forward-ipv6-to-ipv4) which would allow me to stick to IPv4 for internal addresses (I even tried unsuccessfully with `::ffff:10.11.12.13` which is the IPv6 representation of the `10.11.12.13` IPv4 but I just get a "connection refused"). I could use socat for this, or try [Linux implementations of NAT64](http://en.wikipedia.org/wiki/NAT64#Implementations) but that's not the point of this PR, and I find these solutions overly complex so I'd rather expose the containers directly and setup DNS entries, or stick to IPv6 NAT that requires me to have an IPv6 inside containers too.

Thank you all for the feedback.

**Comment by valeriangalliat** [comment]
I amended the "IPv6 NAT" commit to make it optional (and disabled by default), and unify `forwardIPv6Ports` and `forwardPorts` (now we can pass IPv6 destinations to `forwardPorts` if `networking.nat.enableIPv6` is true, it will figure out which ones to pass to iptables or ip6tables).

Also I refactored the code with `mk*` functions to avoid copy/paste; these functions takes `iptables` as parameter (either the string `iptables` or `ip6tables`), and any other parameters that might be different between IPv4 and IPv6.

---

I also tried to setup a public IPv6 inside the container, but I don't know how to do with this virtual ethernet device. Whatever I try, I can't get the container to speak with the outside world without doing NAT. Any idea (maybe @bluescreen303)?

**Comment by valeriangalliat** [comment]
I'm still searching what's the best way to have IPv6 containers. For sure the NAT solution of this PR works fine, and it's in fact the only solution I could get working for now.

Though I finally have found a way to have a publicly routable IPv6 inside containers, so I'm sharing the config to gather some feedback:

```nix

{ config, pkgs, ... }: {
  # Just for IPv4.
  networking.nat.enable = true;
  networking.nat.internalInterfaces = ["ve-+"];
  networking.nat.externalInterface = "enp0s3";

  services.openssh.enable = true;
  services.radvd.enable = true;

  services.radvd.config = ''
    interface ve-web {
      AdvSendAdvert on;
      AdvSourceLLAddress on;

      prefix aaaa:bbbb:cccc:dddd::/64 {
        AdvOnLink on;
        AdvRouterAddr on;
      };
    };
  '';

  containers.web = {
    autoStart = true;
    privateNetwork = true;

    # I still enable IPv4 since it's the only way I found to
    # automatically populate </etc/resolv.conf>, but I need to
    # restart the container after boot to get it, otherwise
    # it's empty.
    hostAddress = "192.168.100.1";
    localAddress = "192.168.100.2";

    config = { config, pkgs, ... }: { };
  };
}
```

And the script I run to enable IPv6:

```sh
# Keep IPv6 autoconfiguration even when enabling forwarding.
sysctl net.ipv6.conf.default.accept_ra=2
sysctl net.ipv6.conf.all.accept_ra=2
sysctl net.ipv6.conf.enp0s3.accept_ra=2

sysctl net.ipv6.conf.default.forwarding=1
sysctl net.ipv6.conf.all.forwarding=1
sysctl net.ipv6.conf.enp0s3.forwarding=1

# Need an IPv6 on container interface.
ip -6 addr add aaaa:bbbb:cccc:dddd::2 dev ve-web

# Container internal IPv6 (autoconfigured thanks to radvd).
ip -6 route add aaaa:bbbb:cccc:dddd:1111:2222:3333:4444 dev ve-web

# Thanks <https://www.ipsidixit.net/2010/03/24/239/>.
sysctl net.ipv6.conf.default.proxy_ndp=1
sysctl net.ipv6.conf.all.proxy_ndp=1
sysctl net.ipv6.conf.enp0s3.proxy_ndp=1

ip -6 neigh add proxy aaaa:bbbb:cccc:dddd:1111:2222:3333:4444 dev enp0s3
ip -6 neigh add proxy aaaa:bbbb:cccc:dddd::1 dev ve-web
```

And the network topology:

```
                 Router
          aaaa:bbbb:cccc:dddd::1
                   |
                 enp0s3
          <autoconfigured ipv6>
                 NixOS
          aaaa:bbbb:cccc:dddd::2
                ve-web
                   |
                 eth0
aaaa:bbbb:cccc:dddd:1111:2222:3333:4444
             Web container
```

Any input on this? The `proxy_ndp` think looks more like a hack to me… and it requires to manually populate the proxy table, however this is currently the only way I found to get public IPv6 inside containers. Have anyone done such a thing before?

---

I also wanted to try with a bridged network (#5857), but I can't find any documentation on how to properly configure a bridge with NixOS… `networking.bridges.br0.interfaces = [ "enp0s3" ];` "works" but I'm loosing IPv4 DHCP (even if I explicitely enable dhcpcd). Concerning IPv6, the container (started with `--network-bridge=br0`, and after running `ip link set host0 up` inside) can ping its host and vice versa, but can't ping any IPv6 address "outside"; the container properly sends neighbour solicitations, and the target replies a proper neighbour advertisment (I can see it from other machines in the network), though the neighbour advertisments don't reach the bridge interface (a tcpdump on NixOS host shows only guest sollicitations but not the advertisment replies, on both `enp0s3` and `br0`), so I'm kinda lost here.

---

In the end I'd like to write a documentation on how to work with IPv6 and NixOS containers (or even integrate this as NixOS configuration options, like for IPv6 NAT), but I need to get this working properly first.

**Comment by ts468** [comment]
@valeriangalliat Thank you very much for your hard work on this PR. I'm very interested in your enhancements, and so I wanted to ask what I could help with to get this merged?

@wkennington What do you think? Any objections, hints or suggestions?

**Comment by edolstra** [comment]
Looks good to me.

**Comment by ts468** [comment]
@edolstra Thanks a lot for your review!

@valeriangalliat Should I rebase your PR onto the current master, or do you have time to maybe update it yourself?

**Comment by wkennington** [comment]
It looks fine, but why parse ipv6 addresses in forward ports instead of
breaking it out like the other options.

**Comment by valeriangalliat** [comment]
@wkennington I'm not sure to understand, do you mean the `isIPv6` check used only in `forwardPorts` and not in, for example, `internalIPs`, `externalIP`?

I added `isIPv6` thanks to @edolstra's comment on Feb 9, but I didn't thought to use it to unify other options... I believe `internalIPs` could benefit from this unification too, because it's also a list of IPs, but options with a single IP like `externalIP`, `hostAddress` and `localAddress`, requiring a singe IP, can only have an IPv6 variant to allow both IPv4 and IPv6 usage.

@ts468 I'm a bit busy these days so I'd really appreciate you rebase this PR, I have no idea it if it's straightforward of if there's some adaptations to do... I gave you commit access to `valeriangalliat/nixpkgs` so you can update this very branch directly. Feel free to ask me if you have any question.

**Comment by valeriangalliat** [comment]
The IPv6 support for containers was added by #14018, and I think the same thing can be achieved with the new bridge support instead of IPv6 NAT. Closing this PR, thanks everyone!

**Comment by ncfavier** [comment]
This would still be useful to me. I am routing all my v4 and v6 traffic through my server via WireGuard, and I'd like to use masquerading for both protocols.

Currently I'm doing something like
```nix
{ # networking.nat
  enable = true;
  externalInterface = "ens3";
  internalIPs = [ "10.42.0.0/16" ];
  extraCommands = "ip6tables -w -t nat -A nixos-nat-post -s fd42::/16 -o ens3 -j MASQUERADE";
}
```
which is not very symmetrical.

Should I open a new issue/PR?

**Pull request closed without merging**